File: agents/livekit/lib/api-client.ts
import logger from "./logger.js";

// Error types used by the LiveKit worker to decide how to signal failures.
// Keep these provider-agnostic so they can be interpreted at higher layers.
export class ApiRequestError extends Error {
  status: number;
  body: any;
  code?: string;
  scope?: string;
  details?: any;

  constructor(status: number, body: any, message: string) {
    super(message);
    this.name = 'ApiRequestError';
    this.status = status;
    this.body = body;
    this.code = body?.code;
    this.scope = body?.scope;
    this.details = body?.details;
  }
}

// Busy-oriented error thrown by call.start() when the agent concurrency limit is exceeded.
// LiveKit uses the thrown error message to map to an intended SIP "busy" cause.
export class AgentConcurrencyLimitExceededBusyError extends Error {
  code = 'AGENT_CONCURRENCY_LIMIT_EXCEEDED';
  status = 429;
  scope?: string;
  details?: any;

  constructor(opts: { scope?: string; details?: any; originalError?: string } = {}) {
    const scopeSuffix = opts.scope ? ` [${opts.scope}]` : '';
    const original = opts.originalError ? ` - ${opts.originalError}` : '';
    super(`busy: AGENT_CONCURRENCY_LIMIT_EXCEEDED${scopeSuffix}${original}`);
    this.name = 'AgentConcurrencyLimitExceededBusyError';
    this.scope = opts.scope;
    this.details = opts.details;
  }
}

// API-related type definitions
export interface Instance {
  id: string;
  metadata?: Record<string, any>;
  Agent?: Agent;
  streamLog?: boolean;
  recording?: {
    enabled: boolean;
    key?: string;
  };
}

export interface Agent {
  id: string;
  userId: string;
  modelName: string;
  organisationId: string;
  prompt?: string;
  options?: {
    /**
     * Optional opening greeting played right after the session starts.
     *
     * Note: for Ultravox realtime, prefer `vendorSpecific.ultravox.firstSpeakerSettings` for
     * provider-native greetings; this block is intended to be portable and primarily used
     * for pipeline and OpenAI realtime sessions.
     */
    greeting?: {
      /**
       * Fixed greeting text.
       * When set, the agent plays this exact line at session start (uninterruptible) and waits for playout.
       */
      text?: string;
      /**
       * LLM instructions for the greeting.
       * When set (and `text` is not set), the agent asks the model to greet the user accordingly.
       */
      instructions?: string;
    };
    /**
     * Optional override for LiveKit voice stack. When omitted, mode is derived from the
     * model id in `modelName` (see GET /models: `voiceStack` / `requiresSttTts`).
     */
    voiceMode?: 'realtime' | 'pipeline';
    /**
     * Optional maximum session duration for realtime LLMs (e.g. "305s").
     * Used by worker when constructing RealtimeModel.
     */
    maxDuration?: string;
    /** Sampling temperature for pipeline LLM (OpenAI / Google plugins). */
    temperature?: number;
    stt?: {
      /**
       * BCP-47 primary tag (e.g. `en`) for pipeline Inference STT.
       * Values like `any` / `multi` are treated as unspecified and default to `en` (or `LIVEKIT_PIPELINE_STT_LANG`).
       */
      language?: string;
      /**
       * STT vendor for LiveKit pipeline (e.g. `deepgram`, `assemblyai`, `cartesia`).
       * You may optionally scope the inference model (and language suffix) via `vendor/model[:lang]`,
       * e.g. `deepgram/nova-3:en` (defaults to language derived from `stt.language`).
       */
      vendor?: string;
    };
    tts?: {
      language?: string;
      /**
       * TTS vendor for LiveKit pipeline (e.g. cartesia, google, elevenlabs).
       * `google` uses Gemini TTS on Node (`@livekit/agents-plugin-google`), not Google Cloud
       * voice ids (`en-GB-Standard-O`). Map timbre with `LIVEKIT_PIPELINE_GEMINI_TTS_VOICE`
       * (global), `LIVEKIT_PIPELINE_GEMINI_TTS_VOICE_<LANG>_<REGION>` (e.g. `..._EN_GB`),
       * or `vendorSpecific.google.geminiVoiceName`. For a custom Inference TTS string, set
       * `LIVEKIT_PIPELINE_GOOGLE_TTS`.
       *
       * You may optionally scope the inference model via `vendor/model`, e.g. `deepgram/aura-2`
       * or `cartesia/sonic-3`. If you include a full `vendor/model:voice` string here, it wins.
       */
      vendor?: string;
      voice?: string;
    };
    /**
     * Optional regular expression pattern to filter outbound calls.
     * The regexp is anchored with ^ and $ to match the complete phone number.
     * Only outbound calls (via originate or transfer) where the destination number matches this pattern will be allowed.
     */
    outboundCallFilter?: string;
    /**
     * Custom prompt to be used by the TransferAgent during consultative transfers.
     * This allows customization of how the TransferAgent introduces the call and interacts with the transfer target.
     * The prompt can include the placeholder ${parentTranscript} which will be replaced with the conversation history.
     * If not specified, a default prompt will be used.
     * This can be overridden on a per-transfer basis by providing transferPrompt as a parameter to the transfer function call.
     */
    transferPrompt?: string;
    /**
     * Fallback configuration for this agent. Used by the LiveKit worker to decide
     * how to recover when the primary model fails to connect or run.
     *
     * Precedence:
     *  1. agent  - restart with a different agent (not yet implemented in worker).
     *  2. model  - restart the session with a different modelName.
     *  3. number - transfer the call to this number using the builtin transfer function.
     */
    fallback?: {
      /**
       * Identifier of an alternative agent to use if the primary agent fails.
       */
      agent?: string;
      /**
       * Fallback model name to use if the primary model fails.
       */
      model?: string;
      /**
       * Fallback transfer destination (phone number or endpoint ID).
       */
      number?: string;
    };
    /**
     * Vendor-specific options for different providers.
     * Each vendor can have its own set of options that are passed through to the provider.
     */
    vendorSpecific?: {
      /**
       * Ultravox-specific options.
       * These options are passed directly to the Ultravox API when creating calls.
       */
      ultravox?: {
        /**
         * Experimental settings for Ultravox calls.
         * Example: { transcriptionProvider: "deepgram-nova-3" } or { transcriptionProvider: "ultravox" }
         */
        experimentalSettings?: {
          transcriptionProvider?: string;
          [key: string]: any;
        };
        /**
         * VAD settings forwarded to Ultravox `POST /api/calls` (`vadSettings`).
         * Durations use protobuf duration strings, e.g. `"0.384s"`.
         * @see https://docs.ultravox.ai/api-reference/calls/calls-post
         */
        vadSettings?: {
          turnEndpointDelay?: string;
          minimumTurnDuration?: string;
          minimumInterruptionDuration?: string;
          frameActivationThreshold?: number;
        };
        /**
         * Opening-turn behaviour (`firstSpeakerSettings`). Prefer this over the model-level
         * deprecated `firstSpeaker` enum; when set here, that enum is not sent on the create-call body.
         * Exactly one of `user` or `agent` should be set per Ultravox API.
         * @see https://docs.ultravox.ai/api-reference/calls/calls-post
         */
        firstSpeakerSettings?: {
          user?: {
            fallback?: {
              delay?: string;
              text?: string;
              prompt?: string;
            };
          };
          agent?: {
            uninterruptible?: boolean;
            text?: string;
            prompt?: string;
            delay?: string;
          };
        };
        /**
         * Messages spoken after cumulative periods of user inactivity (`inactivityMessages`).
         * Durations are protobuf-style strings (e.g. `"30s"`). See Ultravox docs for ordering and `endBehavior`.
         * @see https://docs.ultravox.ai/api-reference/calls/overview#inactivitymessages-5
         */
        inactivityMessages?: Array<{
          duration: string;
          message: string;
          endBehavior?:
            | 'END_BEHAVIOR_UNSPECIFIED'
            | 'END_BEHAVIOR_HANG_UP_SOFT'
            | 'END_BEHAVIOR_HANG_UP_STRICT';
        }>;
        [key: string]: any;
      };
      /**
       * Google / Gemini options for the LiveKit pipeline when using Gemini TTS.
       */
      google?: {
        /**
         * Prebuilt Gemini TTS voice name (e.g. Kore, Puck). Overrides env and Cloud-id defaults.
         */
        geminiVoiceName?: string;
      };
      [key: string]: any;
    };
    /**
     * Recording configuration for this agent.
     * Can be overridden at instance/listener level.
     */
    recording?: {
      enabled: boolean;
      key?: string;
    };
  };
  functions?: AgentFunction[];
  keys?: string[];
}

/**
 * Fetch an agent definition by ID from the main API.
 * Note: this is distinct from the /agent-db/* internal APIs used for calls.
 */
export async function getAgentById(agentId: string): Promise<Agent> {
  return makeApiRequest<Agent>(`/api/agents/${encodeURIComponent(agentId)}`);
}

export interface AgentFunction {
  name: string;
  description: string;
  input_schema: {
    properties: Record<string, {
      type: string;
      required?: boolean;
      [key: string]: any;
    }>;
  };
}

export interface Call {
  id: string;
  parentId?: string;
  userId: string;
  organisationId: string;
  instanceId: string;
  agentId: string;
  platform: string;
  platformCallId?: string;
  calledId?: string;
  callerId?: string;
  modelName?: string;
  options?: any;
  metadata?: {
    aplisay?: {
      callerId?: string;
      calledId?: string;
      fallbackNumbers?: string[];
      model?: string;
      callId?: string;
      [key: string]: any;
    };
    [key: string]: any;
  };
  start(): Promise<void>;
  end(reason?: string): Promise<void>;
}

export interface CallMetadata {
  [key: string]: any;
}

export interface OutboundInfo {
  toNumber: string;
  fromNumber: string;
  aplisayId: string;
  instanceId: string;
}

export interface TrunkInfo {
  id: string;
  name?: string | null;
  outbound: boolean;
  flags?: {
    canRefer?: boolean;
    [key: string]: any;
  } | null;
}

export interface PhoneNumberInfo {
  number: string;
  handler: string;
  instanceId?: string | null;
  organisationId?: string | null;
  outbound?: boolean;
  aplisayId?: string | null;
  trunk?: TrunkInfo | null;
  provisioned?: boolean;
  [key: string]: any;
}

export interface PhoneRegistrationInfo {
  id: string;
  name?: string | null;
  handler: string;
  status?: string;
  state?: string;
  outbound?: boolean;
  organisationId?: string | null;
  instanceId?: string | null;
  registrar?: string | null;
  options?: {
    transport?: string;
    [key: string]: any;
  } | null;
  [key: string]: any;
}

export type PhoneEndpointInfo = PhoneNumberInfo | PhoneRegistrationInfo;

export interface InvocationLogPayload {
  userId: string;
  organisationId: string;
  callId: string;
  subsystem?: string;
  log: any;
}

// Get the API base URL from environment variable
function getApiBaseUrl(): string {
  const serviceBaseUri = process.env.SERVICE_BASE_URI;
  
  if (!serviceBaseUri) {
    throw new Error('SERVICE_BASE_URI environment variable is required');
  }
  
  return serviceBaseUri;
}

// Make an HTTP request to the API
async function makeApiRequest<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  const baseUrl = getApiBaseUrl();
  const url = `${baseUrl}${endpoint}`;
  const sharedToken = process.env.SHARED_API_TOKEN;
  
  //logger.debug({ url, method: options.method || 'GET' }, 'Making API request');
  
  try {
    logger.debug({ url, method: options.method || 'GET', options }, 'Making API request');
    const response = await fetch(url, {
      headers: {
        'Content-Type': 'application/json',
        ...(sharedToken && { 'x-shared-token': sharedToken }),
        ...options.headers,
      },
      ...options,
    });

    if (!response.ok) {
      const errorText = await response.text();
      logger.error(
        {
          url,
          status: response.status,
          statusText: response.statusText,
          error: errorText,
        },
        'API request failed',
      );

      let body: any = null;
      try {
        body = JSON.parse(errorText);
      } catch {
        // If we can't parse JSON, keep a minimal structure.
        body = { raw: errorText };
      }

      throw new ApiRequestError(
        response.status,
        body,
        `API request failed: ${response.status} ${response.statusText}`,
      );
    }

    const data = await response.json();
    //logger.debug({ url, status: response.status }, 'API request successful');
    return data;
  } catch (error) {
    logger.error({ url, error }, 'API request error');
    throw error;
  }
}

// Get instance by ID from the API
export async function getInstanceById(instanceId: string): Promise<any> {
  return makeApiRequest(`/api/agent-db/instance?instanceId=${instanceId}`);
}

// Get instance by phone number from the API
export async function getInstanceByNumber(number: string): Promise<any> {
  return makeApiRequest(`/api/agent-db/instance?number=${encodeURIComponent(number)}`);
}

// Get phone numbers from the API
export async function getPhoneNumbers(handler?: string): Promise<any[]> {
  const query = handler ? `?handler=${encodeURIComponent(handler)}` : '';
  return makeApiRequest(`/api/agent-db/phone-numbers${query}`);
}

// Get phone endpoint by ID (PhoneRegistration)
export async function getPhoneEndpointById(id: string): Promise<PhoneRegistrationInfo | null> {
  try {
    const result = await makeApiRequest<{ items: PhoneRegistrationInfo[] }>(
      `/api/agent-db/phone-endpoints?id=${encodeURIComponent(id)}`
    );
    return result?.items?.[0] || null;
  } catch (error) {
    logger.error({ id, error }, 'Failed to get phone endpoint by id');
    return null;
  }
}

// Get phone endpoint by number (PhoneNumber)
// If trunkId is provided, validates that the call arrived on the correct trunk
export async function getPhoneEndpointByNumber(
  number: string,
  trunkId?: string | null
): Promise<PhoneNumberInfo | null> {
  try {
    let url = `/api/agent-db/phone-endpoints?number=${encodeURIComponent(number)}`;
    if (trunkId) {
      url += `&trunkId=${encodeURIComponent(trunkId)}`;
    }
    const result = await makeApiRequest<{ items: PhoneNumberInfo[] }>(url);
    return result?.items?.[0] || null;
  } catch (error: any) {
    // If it's a trunk mismatch error, re-throw it
    if (error?.message?.includes('Trunk mismatch')) {
      throw error;
    }
    logger.error({ number, trunkId, error }, 'Failed to get phone endpoint by number');
    return null;
  }
}

// Legacy function - kept for backward compatibility, now uses phone-endpoints endpoint
export async function getPhoneNumberByNumber(number: string): Promise<PhoneNumberInfo | null> {
  return getPhoneEndpointByNumber(number);
}

// Mark a phone number as provisioned (or not) in the platform after LiveKit sync
export async function setPhoneNumberProvisioned(
  number: string,
  provisioned: boolean
): Promise<void> {
  try {
    await makeApiRequest<{ success: boolean }>(
      `/api/agent-db/phone-endpoints/${encodeURIComponent(number)}`,
      {
        method: 'PATCH',
        body: JSON.stringify({ provisioned })
      }
    );
  } catch (error) {
    logger.error({ number, provisioned, error }, 'Failed to update phone number provisioning state');
  }
}


// Create a new call record
export async function createCall(callData: {
  id?: string;
  parentId?: string;
  userId: string;
  organisationId: string;
  instanceId: string;
  agentId: string;
  platform: string;
  platformCallId?: string;
  calledId?: string;
  callerId?: string;
  modelName?: string;
  options?: any;
  metadata?: any;
}): Promise<any> {
  const call = await makeApiRequest('/api/agent-db/call', {
    method: 'POST',
    body: JSON.stringify(callData)
  }) as any;
  
  // Add start() and end() methods to the call object
  call.start = async () => {
    logger.debug({ call }, "logging starting call");
    try {
      return await makeApiRequest(`/api/agent-db/call/${call.id}/start`, {
        method: 'POST',
        body: JSON.stringify({
          userId: call.userId,
          organisationId: call.organisationId
        })
      });
    } catch (err) {
      if (
        err instanceof ApiRequestError &&
        err.status === 429 &&
        err.code === 'AGENT_CONCURRENCY_LIMIT_EXCEEDED'
      ) {
        throw new AgentConcurrencyLimitExceededBusyError({
          scope: err.scope,
          details: err.details,
          originalError: err.body?.error ?? err.body?.raw,
        });
      }
      throw err;
    }
  };
  
  call.end = async (reason?: string, transactionLogs?: Array<{
    userId: string;
    organisationId: string;
    callId: string;
    type: string;
    data?: string;
    isFinal?: boolean;
    createdAt?: Date;
  }>) => {
    // Make this function idempotent - if already called, return the existing promise
    if ((call as any)._endCalled) {
      logger.debug({ callId: call.id, reason }, "call.end() already called, returning existing promise");
      return (call as any)._endPromise;
    }

    // Mark as called and store the promise
    (call as any)._endCalled = true;
    
    logger.debug({ call, reason, transactionLogCount: transactionLogs?.length }, "logging ending call");
    const body: any = { 
      reason,
      userId: call.userId,
      organisationId: call.organisationId
    };
    // Use provided transactionLogs, or fall back to batched logs on call object
    const logsToSend = transactionLogs || (call as any).batchedTransactionLogs;
    if (logsToSend && logsToSend.length > 0) {
      // Convert Date objects to ISO strings for JSON serialization
      body.transactionLogs = logsToSend.map((log: any) => ({
        ...log,
        createdAt: log.createdAt instanceof Date 
          ? log.createdAt.toISOString() 
          : log.createdAt ? new Date(log.createdAt).toISOString() : undefined
      }));
    }
    
    // Create and store the promise
    const endPromise = makeApiRequest(`/api/agent-db/call/${call.id}/end`, {
      method: 'POST',
      body: JSON.stringify(body)
    });
    
    (call as any)._endPromise = endPromise;
    
    // Handle errors to ensure the promise is still stored even on failure
    return endPromise.catch((error) => {
      logger.error({ callId: call.id, error }, "error in call.end(), but keeping promise for idempotency");
      throw error;
    });
  };
  
  return call;
}

export async function endCallById(callId: string, reason?: string): Promise<any> {
  return makeApiRequest(`/api/agent-db/call/${callId}/end`, {
    method: 'POST',
    body: JSON.stringify({ reason })
  });
}

export async function setCallRecordingData(
  callId: string,
  recordingId: string,
  encryptionKey?: string,
): Promise<void> {
  const body: any = { recordingId };
  if (encryptionKey) {
    body.encryptionKey = encryptionKey;
  }
  await makeApiRequest(`/api/agent-db/call/${encodeURIComponent(callId)}/recording`, {
    method: 'PUT',
    body: JSON.stringify(body),
  });
}

// Create a new transaction log record
export async function createTransactionLog(transactionData: {
  userId: string;
  organisationId: string;
  callId: string;
  type: string;
  data?: string;
  isFinal?: boolean;
  createdAt?: Date;
}): Promise<any> {
  // we don't log status change events
  if (transactionData.type === 'status') {
    return null;
  }
  
  // Prepare the request body, converting Date to ISO string if present
  const requestBody: any = { ...transactionData };
  if (requestBody.createdAt instanceof Date) {
    requestBody.createdAt = requestBody.createdAt.toISOString();
  } else if (requestBody.createdAt) {
    // If it's already a string or other format, ensure it's a valid ISO string
    requestBody.createdAt = new Date(requestBody.createdAt).toISOString();
  }
  
  return makeApiRequest('/api/agent-db/transaction-log', {
    method: 'POST',
    body: JSON.stringify(requestBody)
  });
}

// Create a new invocation log record (compressed and stored server-side)
export async function saveInvocationLog(payload: InvocationLogPayload): Promise<any> {
  return makeApiRequest('/api/agent-db/invocation-log', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}



export { getApiBaseUrl, makeApiRequest };

File: agents/pipecat/pipecat_aplisay/api_client.py
"""REST callbacks back into the llm-agent agent-db API.

Section 8 of docs/livekit-agent-architecture.md is authoritative. Endpoints,
header names, and idempotency behaviour mirror the LiveKit worker's
``api-client.ts`` so the same REST server can serve both implementations.
"""

from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any, Optional

import httpx
from loguru import logger
from pydantic import BaseModel


class ApiRequestError(Exception):
    def __init__(self, status: int, body: Any, message: str) -> None:
        super().__init__(message)
        self.status = status
        self.body = body or {}
        self.code = self.body.get("code") if isinstance(self.body, dict) else None
        self.scope = self.body.get("scope") if isinstance(self.body, dict) else None
        self.details = self.body.get("details") if isinstance(self.body, dict) else None


class AgentConcurrencyLimitExceededBusyError(Exception):
    code = "AGENT_CONCURRENCY_LIMIT_EXCEEDED"
    status = 429

    def __init__(
        self,
        scope: Optional[str] = None,
        details: Any = None,
        original_error: Optional[str] = None,
    ) -> None:
        scope_suffix = f" [{scope}]" if scope else ""
        original = f" - {original_error}" if original_error else ""
        super().__init__(f"busy: AGENT_CONCURRENCY_LIMIT_EXCEEDED{scope_suffix}{original}")
        self.scope = scope
        self.details = details


def _base_url() -> str:
    base = os.environ.get("SERVICE_BASE_URI")
    if not base:
        raise RuntimeError("SERVICE_BASE_URI environment variable is required")
    return base


def _shared_token() -> Optional[str]:
    return os.environ.get("SHARED_API_TOKEN")


async def _request(
    method: str,
    endpoint: str,
    *,
    params: Optional[dict] = None,
    body: Any = None,
    timeout: float = 30.0,
) -> Any:
    url = f"{_base_url()}{endpoint}"
    headers = {"Content-Type": "application/json"}
    token = _shared_token()
    if token:
        headers["x-shared-token"] = token

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.request(method, url, params=params, json=body, headers=headers)
    except httpx.ConnectError as e:
        # DNS / TCP failure reaching the llm-agent REST server. Most common
        # cause in dev is SERVICE_BASE_URI unset, pointed at an unresolvable
        # placeholder, or set to https:// against a plain-HTTP server (a
        # WRONG_VERSION_NUMBER error surfaces as ConnectError too).
        msg = (
            f"cannot reach llm-agent REST server at {url}: {e}. "
            "Check SERVICE_BASE_URI (scheme + host) on the worker and that "
            "the llm-agent server is running and reachable from this host."
        )
        logger.bind(url=url, method=method, error=str(e)).error(msg)
        raise ApiRequestError(502, {"error": msg}, msg) from e
    except httpx.RequestError as e:
        # Other transport-level failures (timeout, TLS handshake errors, etc).
        msg = f"transport error calling {method} {url}: {e}"
        logger.bind(url=url, method=method, error=str(e)).error(msg)
        raise ApiRequestError(502, {"error": msg}, msg) from e

    if resp.status_code >= 400:
        text = resp.text
        try:
            parsed = resp.json()
        except Exception:  # noqa: BLE001
            parsed = {"raw": text}

        log_fn = (
            logger.info if resp.status_code == 404 else logger.warning if resp.status_code < 500 else logger.error
        )
        log_fn(
            f"API request failed: {method} {url} -> {resp.status_code} {text[:200]}"
        )
        raise ApiRequestError(resp.status_code, parsed, f"API request failed: {resp.status_code}")

    if resp.headers.get("content-type", "").startswith("application/json"):
        return resp.json()
    return resp.text


# ---- Lookup ----


def pop_organisation_keys(doc: Any) -> dict:
    """Remove and return the BYOK provider-key bag from a fetched document.

    The agent-db API delivers the organisation's decrypted provider keys
    per call as a top-level ``organisationKeys`` object — a sibling of
    ``Agent`` on the instance document, and a property of the agent json
    from ``/api/agent-db/agent`` (see docs/byok.md). Callers pop it off
    immediately after the fetch so the instance / agent dicts they store,
    serialise, or log never carry key material. Returns ``{}`` when the
    document has no bag (the server omits it when empty).
    """
    if not isinstance(doc, dict):
        return {}
    bag = doc.pop("organisationKeys", None)
    return bag if isinstance(bag, dict) else {}


async def get_instance_by_id(instance_id: str) -> dict:
    # The returned document may carry a top-level ``organisationKeys`` bag
    # (BYOK); callers that build sessions pop it via pop_organisation_keys.
    return await _request("GET", "/api/agent-db/instance", params={"instanceId": instance_id})


async def get_instance_by_number(number: str) -> dict:
    return await _request("GET", "/api/agent-db/instance", params={"number": number})


async def get_agent_by_id(agent_id: str) -> dict:
    return await _request("GET", f"/api/agents/{agent_id}")


async def get_internal_agent_by_id(
    agent_id: str, expected_organisation_id: Optional[str] = None
) -> dict:
    """Fetch a full agent definition (including keys) via the internal agent-db API.

    Used for in-call ``transfer_agent`` handover. Always pass the calling
    call's organisation id so the server can refuse cross-tenant fetches —
    mirrors ``getInternalAgentById`` in the LiveKit worker's api-client.ts.

    The returned agent json may carry an ``organisationKeys`` property (the
    incoming agent's own BYOK bag) — callers must pop it off with
    :func:`pop_organisation_keys` before storing or embedding the dict.
    """
    params: dict = {"agentId": agent_id}
    if expected_organisation_id:
        params["expectedOrganisationId"] = expected_organisation_id
    return await _request("GET", "/api/agent-db/agent", params=params)


async def invoke_subagent(
    agent_id: str,
    input_args: dict,
    metadata: Optional[dict],
    *,
    organisation_id: str,
    call_id: Optional[str] = None,
) -> Any:
    """Invoke a ``text`` type agent as a subagent via the internal agent-db API.

    Returns the subagent's result payload (the arguments it passed to its
    builtin ``result`` function). The server bounds the invocation with its
    own SUBAGENT_TIMEOUT (default 60s), so the HTTP timeout here is set a
    little above that.
    """
    data = await _request(
        "POST",
        "/api/agent-db/subagent",
        body={
            "agentId": agent_id,
            "input": input_args,
            "metadata": metadata,
            "organisationId": organisation_id,
            "callId": call_id,
        },
        timeout=75.0,
    )
    return (data or {}).get("result") if isinstance(data, dict) else data


async def authorise_outbound_destination(
    *,
    called_id: str,
    caller_id: Optional[str] = None,
    agent_options: Optional[dict] = None,
    organisation_id: Optional[str] = None,
    user_id: Optional[str] = None,
    aplisay_id: Optional[str] = None,
    outbound_trunk_id: Optional[str] = None,
    registration_originated: bool = False,
) -> dict:
    """Authorise one outbound destination against the platform's policy.

    Returns the decision dict (``allowed``, ``code``, ``reason``, ``chargeable``,
    ``trunkId``, ``destination``). Raises on transport/server failure — the caller
    (``outbound_filter.authorise_destination``) treats that as a REFUSAL, since the
    policy it enforces (per-trunk operator filter + destination rating on our own
    carrier trunks) cannot be evaluated here.
    """
    return await _request(
        "POST",
        "/api/agent-db/outbound-authorisation",
        body={
            "calledId": called_id,
            "callerId": caller_id,
            "agentOptions": agent_options or {},
            "organisationId": organisation_id,
            "userId": user_id,
            "aplisayId": aplisay_id,
            "outboundTrunkId": outbound_trunk_id,
            "registrationOriginated": bool(registration_originated),
        },
    )


async def get_phone_number(number: str) -> Optional[dict]:
    """Look up a provisioned PhoneNumber (DDI/trunk number) by E.164 number.

    Returns the row (``number``, ``outbound``, ``aplisayId``, ``instanceId``,
    ...) or ``None`` if not found. Used to resolve the egress trunk + validate
    the caller-ID for a WebRTC-origin transfer — mirrors the PhoneNumber lookup
    in LiveKit's ``transfer-handler.ts`` ``validateAndResolveCallerId``.
    """
    try:
        result = await _request(
            "GET", "/api/agent-db/phone-numbers", params={"number": number}
        )
    except ApiRequestError:
        return None
    # Route returns a bare array of rows.
    if isinstance(result, list):
        return result[0] if result else None
    items = result.get("items") if isinstance(result, dict) else None
    return items[0] if items else None


async def get_phone_endpoint_by_id(endpoint_id: str) -> Optional[dict]:
    try:
        result = await _request("GET", "/api/agent-db/phone-endpoints", params={"id": endpoint_id})
    except ApiRequestError:
        return None
    items = result.get("items") if isinstance(result, dict) else None
    return items[0] if items else None


async def get_phone_endpoint_by_number(
    number: str, trunk_id: Optional[str] = None
) -> Optional[dict]:
    params: dict = {"number": number}
    if trunk_id:
        params["trunkId"] = trunk_id
    try:
        result = await _request("GET", "/api/agent-db/phone-endpoints", params=params)
    except ApiRequestError as e:
        if "Trunk mismatch" in str(e):
            raise
        return None
    items = result.get("items") if isinstance(result, dict) else None
    return items[0] if items else None


async def set_phone_endpoint_provisioned(number: str, provisioned: bool) -> None:
    try:
        await _request(
            "PATCH",
            f"/api/agent-db/phone-endpoints/{number}",
            body={"provisioned": provisioned},
        )
    except ApiRequestError as e:
        logger.error(f"failed to update phone provisioning state: {e}")


# ---- Call lifecycle ----


class CallRecord(BaseModel):
    id: str
    userId: str
    organisationId: str
    instanceId: str
    agentId: str
    metadata: dict = {}
    options: dict = {}

    # Set False for browser / WebRTC sessions where the worker fabricates a
    # stub Call locally instead of POSTing /api/agent-db/call. end_call() is a
    # no-op for non-persisted records.
    persisted: bool = True

    # Internal — accumulates batched logs when streamLog is False; flushed at end.
    batched_transaction_logs: list = []
    end_called: bool = False


async def create_call(call_data: dict) -> CallRecord:
    raw = await _request("POST", "/api/agent-db/call", body=call_data)
    record = CallRecord(
        id=raw["id"],
        userId=raw["userId"],
        organisationId=raw["organisationId"],
        instanceId=raw["instanceId"],
        agentId=raw["agentId"],
        metadata=raw.get("metadata") or {},
        options=raw.get("options") or {},
    )
    return record


async def start_call(call: CallRecord) -> None:
    try:
        await _request(
            "POST",
            f"/api/agent-db/call/{call.id}/start",
            body={"userId": call.userId, "organisationId": call.organisationId},
        )
    except ApiRequestError as err:
        if err.status == 429 and err.code == "AGENT_CONCURRENCY_LIMIT_EXCEEDED":
            raise AgentConcurrencyLimitExceededBusyError(
                scope=err.scope,
                details=err.details,
                original_error=(err.body or {}).get("error") if isinstance(err.body, dict) else None,
            )
        raise


async def end_call(call: CallRecord, reason: Optional[str] = None) -> None:
    if call.end_called:
        return
    call.end_called = True
    # Non-persisted stub records (browser sessions) were never POSTed to the
    # server; there's nothing for /call/:id/end to find. Calling it would 404.
    if not call.persisted:
        logger.debug(f"end_call: skipping non-persisted call {call.id}")
        return
    body: dict = {
        "reason": reason,
        "userId": call.userId,
        "organisationId": call.organisationId,
    }
    if call.batched_transaction_logs:
        body["transactionLogs"] = [
            _serialise_log(entry) for entry in call.batched_transaction_logs
        ]
    try:
        await _request("POST", f"/api/agent-db/call/{call.id}/end", body=body)
    except ApiRequestError as e:
        logger.error(f"call.end failed: {e}")
        raise


async def end_call_by_id(call_id: str, reason: Optional[str] = None) -> None:
    await _request("POST", f"/api/agent-db/call/{call_id}/end", body={"reason": reason})


# ---- Logging ----


def _serialise_log(entry: dict) -> dict:
    out = dict(entry)
    created = out.get("createdAt")
    if isinstance(created, datetime):
        out["createdAt"] = created.astimezone(timezone.utc).isoformat()
    return out


async def create_transaction_log(entry: dict) -> None:
    if entry.get("type") == "status":
        return
    body = _serialise_log(entry)
    await _request("POST", "/api/agent-db/transaction-log", body=body)


async def save_invocation_log(payload: dict) -> None:
    await _request("POST", "/api/agent-db/invocation-log", body=payload)


async def save_usage(records: Any) -> None:
    """Post one or more usage meters (LLM tokens, TTS characters, STT audio, …)
    to the platform usage ledger. Accepts a single record dict or a list."""
    body = {"records": records} if isinstance(records, list) else records
    await _request("POST", "/api/agent-db/usage", body=body)


# ---- Recording metadata ----


async def set_call_recording_data(
    call_id: str, recording_id: str, encryption_key: Optional[str] = None
) -> None:
    body: dict = {"recordingId": recording_id}
    if encryption_key:
        body["encryptionKey"] = encryption_key
    await _request("PUT", f"/api/agent-db/call/{call_id}/recording", body=body)
